dev-tools/support-bundle-reader-lib/src/bundle_accessor.rs

```diff
@@ -0,0 +1,260 @@
+        Self { client, id }
+    }
+}
+
+// Access for: The nexus internal API
+#[async_trait]
+impl<'c> SupportBundleAccessor for InternalApiAccess<'c> {
+    async fn get_index(&self) -> Result<SupportBundleIndex> {
+        let stream = self
+            .client
+            .support_bundle_index(self.id.as_untyped_uuid())
+            .await
+            .with_context(|| {
+                format!("downloading support bundle index {}", self.id)
+            })?
+            .into_inner_stream();
+        let s = utf8_stream_to_string(stream).await?;
+
+        Ok(SupportBundleIndex::new(&s))
+    }
+
+    async fn get_file<'a>(
+        &mut self,
+        path: &Utf8Path,
+    ) -> Result<BoxedFileAccessor<'a>>
+    where
+        'c: 'a,
+    {
+        let mut file =
+            StreamedFile::new(self.client, self.id, path.to_path_buf());
+        file.start_stream()
+            .await
+            .with_context(|| "failed to start stream in get_file")?;
+        Ok(Box::new(file))
+    }
+}
+
+pub struct LocalFileAccess {
+    archive: zip::read::ZipArchive<std::fs::File>,
+}
+
+impl LocalFileAccess {
+    pub fn new(path: &Utf8Path) -> Result<Self> {
+        let file = std::fs::File::open(path)?;
+        Ok(Self { archive: zip::read::ZipArchive::new(file)? })
+    }
+}
+
+// Access for: Local zip files
+#[async_trait]
+impl SupportBundleAccessor for LocalFileAccess {
+    async fn get_index(&self) -> Result<SupportBundleIndex> {
+        let names: Vec<&str> = self.archive.file_names().collect();
+        let all_names = names.join("\n");
+        Ok(SupportBundleIndex::new(&all_names))
+    }
+
+    async fn get_file<'a>(
+        &mut self,
+        path: &Utf8Path,
+    ) -> Result<BoxedFileAccessor<'a>> {
+        let mut file = self.archive.by_name(path.as_str())?;
+        let mut buf = Vec::new();
+        std::io::copy(&mut file, &mut buf)?;
+
+        Ok(Box::new(AsyncZipFile { buf, copied: 0 }))
+    }
+}
+
+// We're currently buffering the entire file into memory, mostly because dealing with the lifetime
+// of ZipArchive and ZipFile objects is so difficult.
+pub struct AsyncZipFile {
+    buf: Vec<u8>,
+    copied: usize,
+}
+
+impl AsyncRead for AsyncZipFile {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let to_copy =
+            std::cmp::min(self.buf.len() - self.copied, buf.remaining());
+        if to_copy == 0 {
+            return Poll::Ready(Ok(()));
+        }
+        let src = &self.buf[self.copied..];
+        buf.put_slice(&src[..to_copy]);
+        self.copied += to_copy;
+        Poll::Ready(Ok(()))
+    }
+}
+
+async fn utf8_stream_to_string(
+    mut stream: impl futures::Stream<Item = reqwest::Result<bytes::Bytes>>
+    + std::marker::Unpin,
+) -> Result<String> {
+    let mut result = String::new();
+
+    // When we read from the string, we might not read a whole UTF-8 sequence.
```

Comment:
Why don't we just collect the stream into a `Vec<u8>` and then attempt to to call `String::from_utf8`?